android/app/src/main/java/org/chrome/device/ml/ChromeActivity.java

```diff
@@ -0,0 +1,235 @@
+        }
+      }
+    };
+
+    serviceCallback = new RemoteServiceCallback.Stub() {
+      // This is called by the remote service regularly to tell us about new values
+      public void timeChanged(double time) {
+        Message msg = new Message();
+        msg.what = MSG_TIME_UPDATE;
+        msg.obj = time;
+        serviceHandler.sendMessage(msg);
+      }
+    };
+
+    bindService(mBindIntent, this, Context.BIND_AUTO_CREATE);
+  }
+
+  @Override
+  protected void onStop() {
+    super.onStop();
+    if (mService != null) {
+      try {
+        mService.unregisterCallback(serviceCallback);
+      } catch (RemoteException e) {
+        Log.e(TAG, "Error unregister callback");
+      }
+    }
+    unbindService(this);
+    stopService(new Intent(ChromeActivity.this, MLService.class));
+
+  }
+
+  @Override
+  public boolean onCreateOptionsMenu(Menu menu) {
+    // Inflate the menu; this adds items to the action bar if it is present
+    getMenuInflater().inflate(R.menu.menu_main, menu);
+    return true;
+  }
+
+  @Override
+  public boolean onOptionsItemSelected(MenuItem item) {
+    // Handle action bar item clicks here. The action bar will
+    // automatically handle clicks on the Home/Up button, so long
+    // as you specify a parent activity in AndroidManifest.xml.
+    int id = item.getItemId();
+
+    // noinspection SimplifiableIfStatement
+    if (id == R.id.action_settings) {
+      return true;
+    }
+
+    return super.onOptionsItemSelected(item);
+  }
+
+  public void addListenerOnSpinnerItemSelection() {
+    modelSpinner.setOnItemSelectedListener(new AdapterView.OnItemSelectedListener() {
+      @Override
+      public void onItemSelected(AdapterView<?> adapterView, View view, int i, long l) {
+        spinnerSelection = i;
+        textboxAppend("Model Selected: " + MODELS[spinnerSelection] + "\n");
+      }
+
+      @Override
+      public void onNothingSelected(AdapterView<?> adapterView) {
+        spinnerSelection = 0;
+      }
+    });
+  }
+
+  private void addItemsOnSpinner() {
+    List<String> list = new ArrayList<String>();
+    for (int i=0; i<MODELS.length; i++) {
+      list.add(MODELS[i]);
+    }
+    ArrayAdapter<String> dataAdapter = new ArrayAdapter<String>(getApplicationContext(),
+      android.R.layout.simple_spinner_item, list);
+    dataAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
+    modelSpinner.setAdapter(dataAdapter);
+  }
+
+  // Handles button actions
+  private void buttonHandler() {
+    textboxAppend("IPC\n");
+    Log.v(TAG, "IPC");
+
+    this.startService(mBindIntent);
+  }
+
+  @Override
+  public void onServiceConnected(ComponentName componentName, IBinder service) {
+    Log.v(TAG, "Service connected.");
+    mService = RemoteService.Stub.asInterface(service);
+
+    // Monitor service
+    try{
+      mService.registerCallback(serviceCallback);
+    } catch (RemoteException e) {
+      Log.e(TAG, e.getStackTrace().toString());
```

Comment:
would be useful to also include that this was an error when registering the callback